### PR TITLE
pg_resgroup_move_query() improvements.

### PR DESCRIPTION
**Pull request by SmartKeyerror**
This is the copy of the PR #13189, since I could not change the original PR content, have to create a new one.

----

The pg_resgroup_move_query() function implemented in 51ee26b has several disadvantages:
1) Slot leaking. The slot acquired in initiator process doesn't free if target process isn't alive, didn't receive a signal, or even received, but at the time it was in idle state.
2) Race condition between UnassignResGroup() called at the end of transaction (or honestly not only UnassignResGroup(), but any other code) and handler called immediately after USR1 signal.
3) Not working entrydb process moving. Previously, the USR1 signal was sent to only one process, target or entrydb. Entrydb moving was never tested properly.

Improvements added to solve the first problem:
1) Feedback from target process to initiator. Target process can set initiator's latch to quickly interrupt pg_resgroup_move_query() from awaiting.
2) Existed movetoResSlot parameter acts like a mark. Initiator sets it and waits on latch. If movetoResSlot become NULL, slot control is on the target process side.
3) Initiator PID. Used by target process to get initiator's latch.
4) Mutex. To guard all critical moveto* parameters from parallel changes.

To solve the second problem, there was an attempt to use Postgres interruptions. I was unhappy to know GPDB use raw InterruptPending value to do some pre-cancellation of dispatched queries. GPDB thinks InterruptPending can be triggered only by "negative" events, which leads to cancellation. The temporary solution with InterruptPending-like "positive" flag showed that we may wait for next CHECK_FOR_INTERRUPTS() call for a long. For example, some fat insert may do group moving only at the end of dispatching, which makes no sense.
Thus, decreasing the probability of races is done now by additional IsTransactionState() check.

IMPORTANT NOTICE. The existed solution still affected by race conditions. The current handler's implementation is an example of bad design and should be reworked (moved to CHECK_FOR_INTERRUPTS()).
(This part is related to [this](https://github.com/greenplum-db/gpdb/pull/13148). We decided to do a separate patch later - current PR is already big and fixes a problem in most cases.)

To solve the third problem, now we send signals to target and entrydb processes separately. We do UnassignResGroup() for one process only, not cleaning all counters for still working entrydb process inside handler. More, entrydb moving logic is now separated from segments moving, and so, it became more readable.

New GUC (gp_resource_group_move_timeout) was added to limit a time we're waiting for feedback from target.

New regression tests shows there is no slot leaking with some rare and now correctly resolved cases.

### CODE REVIEW

**Comment by yaowangm** [comment]
@SmartKeyerror It seems there are still some code conflicts. When I working on the review, could you please solve the conflcts asap so we can put the review on a solid base? Thanks.

**Comment by SmartKeyerror** [comment]
> @SmartKeyerror It seems there are still some code conflicts. When I working on the review, could you please solve the conflcts asap so we can put the review on a solid base? Thanks.

The conflict is not about move query,

**Review by yaowangm — COMMENTED** [review]
As talked, please remember to remove the useless empty files before delivering.

File: src/backend/utils/resgroup/resgroup.c [hunk 1]
```diff
@@ -3297,28 +3433,112 @@ HandleMoveResourceGroup(void)
...
+/*
+ * Try to give away all slot control to target process.
+ */
+static void
+resGroupGiveSlotAway(int sessionId, ResGroupSlotData ** slot, Oid groupId)
+{
+	long		timeout;
+	int64		curTime;
+	int64		waitStart;
+	int			latchRes;
```
[the remark this answers is not in the document]
Comment (SmartKeyerror) [reply src/backend/utils/resgroup/resgroup.c:3445]: I think it's the display bug on GitHub.

File: src/include/storage/proc.h [hunk 2]
```diff
@@ -224,8 +224,18 @@ struct PGPROC
 	void		*resSlot;	/* the resource group slot granted.
 							 * NULL indicates the resource group is
 							 * locked for drop. */
-	void		*movetoResSlot; /* the resource group slot move to, valid only on QD */
-	Oid			movetoGroupId;  /* the resource group id move to */
+	slock_t		movetoMutex; /* spinlock to protect moveto* fields below */
+	void		*movetoResSlot; /* the resource group slot move to, valid only
```
[the remark this answers is not in the document]
Comment (SmartKeyerror) [reply src/include/storage/proc.h:228]: Have no idea, maybe need to ask the original author.

File: src/test/isolation2/sql/resgroup/resgroup_move_query.sql [hunk 3]
```diff
@@ -43,5 +52,179 @@ SELECT is_session_in_group(pid, 'default_group') FROM pg_stat_activity WHERE wai
 2<:
 2: END;
 
+-- test3: the destination group will wake up 'pg_resgroup_move_query' when a new slot become available
+1: SET ROLE role_move_query;
+1&: SELECT pg_sleep(5);
```
[the remark this answers is not in the document]
Comment (SmartKeyerror) [reply src/test/isolation2/sql/resgroup/resgroup_move_query.sql:57]: Thanks for your recommendation!
The waiting condition needs more than one process to wait for the same advisory lock, so it needs about 4 sessions to accomplish the test, it's a little complicated.

File: src/backend/storage/ipc/procarray.c [hunk 4]
```diff
@@ -5277,6 +5277,10 @@ ResGroupMoveCheckTargetReady(int sessionId, bool *clean, bool *result)
 		backendId = proc->backendId;
 
 		SpinLockAcquire(&proc->movetoMutex);
+		/* If proc->movetoCallerPid not equals to MyProc->pid, the target
+		 * process could is handling signal from another caller.After we
```
Comment (yaowangm) [inline src/backend/storage/ipc/procarray.c:5281]: Please add a space before "After we..."

**Review by yaowangm — APPROVED** [review]
Thanks for your work. As talked, I will approve the PR based on current change and please determine the rest by yourself.

**Comment by yaowangm** [comment]
Please remember to backport it to 6X and refer the PR here.

**Pull request merged**